### ISSUE TITLE
Efficiency considerations in CustomNonbondedForce.addInteractionGroup?

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#1765]
I've been tinkering with [alternative PME splitting schemes](http://www.columbia.edu/cu/chemistry/groups/berne/papers/jcp_115_2348_2001.pdf) that split shirt-range real interactions into a separate force group by adding and subtracting `CustomNonbondedForce`s to different force groups, as well as [solute-solvent splitting schemes](http://rspa.royalsocietypublishing.org/content/472/2189/20160138) that restrict the interactions included in the inner force groups to solute-solute and solute-solvent interactions.

Surprisingly, restricting the `CustomNonbondedForce` terms with
```python
force.addInteractionGroup(solute_atoms, all_atoms)
```
where `solute_atoms` are ~5,000 atoms (packed to the beginning of the system) and `all_atoms` are ~50,000 atoms actually increases the execution time (see ns/day on right of following output):
```
SrcExplicit              | solute_solvent_splitting False | constraint_tol 1.000e-08 | ewald_tol 1.000e-04 | cutoff   10.000 A | HMR     True | timestep    2.000 fs | nvalence  4 | nshort  2 | nmedium  2 |     24.4 % |      8.1 ns/day
SrcExplicit              | solute_solvent_splitting  True | constraint_tol 1.000e-08 | ewald_tol 1.000e-04 | cutoff   10.000 A | HMR     True | timestep    2.000 fs | nvalence  4 | nshort  2 | nmedium  2 |     60.3 % |      2.4 ns/day
```
What efficiency considerations should we be aware of when using `addInteractionGroup`?
```python
from simtk import openmm, unit
from simtk.openmm import app
import numpy as np
import copy
import time
import os, os.path
import sys

nsteps = 1000

mpicomm = None

kB = unit.BOLTZMANN_CONSTANT_kB * unit.AVOGADRO_CONSTANT_NA

class MTSGHMCIntegrator(openmm.CustomIntegrator):

    """
    MTS Generalized hybrid Monte Carlo (GHMC) integrator with 4 hardcoded force groups (for experiments only).

    """

    def __init__(self, temperature=298.0 * unit.kelvin, collision_rate=91.0 / unit.picoseconds, timestep=1.0 * unit.femtoseconds, nvalence=1, nshort=1, nmedium=1):
        """
        Create a generalized hybrid Monte Carlo (GHMC) integrator.

        Parameters
        ----------
        temperature : unit.Quantity compatible with kelvin, default: 298*unit.kelvin
           The temperature.
        collision_rate : unit.Quantity compatible with 1/picoseconds, default: 91.0/unit.picoseconds
           The collision rate.
        timestep : unit.Quantity compatible with femtoseconds, default: 1.0*unit.femtoseconds
           The integration timestep.

        Notes
        -----
        This integrator is equivalent to a Langevin integrator in the velocity Verlet discretization with a
        Metrpolization step to ensure sampling from the appropriate distribution.

        Additional global variables 'ntrials' and  'naccept' keep track of how many trials have been attempted and
        accepted, respectively.

        References
        ----------
        Lelievre T, Stoltz G, and Rousset M. Free Energy Computations: A Mathematical Perspective
        http://www.amazon.com/Free-Energy-Computations-Mathematical-Perspective/dp/1848162472

        """

        # Initialize constants.
        kT = kB * temperature
        gamma = collision_rate

        # Create a new custom integrator.
        super(MTSGHMCIntegrator, self).__init__(timestep)

        #
        # Integrator initialization.
        #
        self.addGlobalVariable("kT", kT)  # thermal energy
        self.addGlobalVariable("b", np.exp(-gamma * timestep))  # velocity mixing parameter
        self.addPerDofVariable("sigma", 0) # velocity standard deviation
        self.addGlobalVariable("ke", 0)  # kinetic energy
        self.addPerDofVariable("vold", 0)  # old velocities
        self.addPerDofVariable("xold", 0)  # old positions
        self.addGlobalVariable("Eold", 0)  # old energy
        self.addGlobalVariable("Enew", 0)  # new energy
        self.addGlobalVariable("potential_old", 0)  # old potential energy
        self.addGlobalVariable("potential_new", 0)  # new potential energy
        self.addGlobalVariable("accept", 0)  # accept or reject
        self.addGlobalVariable("naccept", 0)  # number accepted
        self.addGlobalVariable("ntrials", 0)  # number of Metropolization trials
        self.addPerDofVariable("x1", 0)  # position before application of constraints
        #
        # Initialization.
        #
        self.beginIfBlock("ntrials = 0")
        self.addComputePerDof("sigma", "sqrt(kT/m)")
        self.addConstrainPositions()
        self.addConstrainVelocities()
        self.endBlock()

        #
        # Allow context updating here.
        #
        self.addUpdateContextState()

        #
        # Velocity randomization
        #
        self.addComputePerDof("v", "sqrt(b)*v + sqrt(1-b)*sigma*gaussian")
        self.addConstrainVelocities()

        # Compute initial total energy
        self.addComputeSum("ke", "0.5*m*v*v")
        self.addComputeGlobal("potential_old", "energy")
        self.addComputeGlobal("Eold", "ke + potential_old")
        self.addComputePerDof("xold", "x")
        self.addComputePerDof("vold", "v")
        # RESPA MTS step
        self.addComputePerDof("v", "v + 0.5*dt*f3/m")
        for medium in range(nmedium):
            self.addComputePerDof("v", "v + 0.5*(dt/%d)*f2/m" % nmedium)
            for short in range(nshort):
                self.addComputePerDof("v", "v + 0.5*(dt/%d)*f1/m" % (nmedium*nshort))
                for valence in range(nvalence):
                    self.addComputePerDof("v", "v + 0.5*(dt/%d)*f0/m" % (nmedium*nshort*nvalence))
                    self.addComputePerDof("x", "x + v*(dt/%d)" % (nmedium*nshort*nvalence))
                    self.addComputePerDof("x1", "x")
                    self.addConstrainPositions()
                    self.addComputePerDof("v", "v + (x-x1)/(dt/%d)" % (nmedium*nshort*nvalence))
                    self.addConstrainVelocities()
                    self.addComputePerDof("v", "v + 0.5*(dt/%d)*f0/m" % (nmedium*nshort*nvalence))
                self.addComputePerDof("v", "v + 0.5*(dt/%d)*f1/m" % (nmedium*nshort))
            self.addComputePerDof("v", "v + 0.5*(dt/%d)*f2/m" % nmedium)
        self.addComputePerDof("v", "v + 0.5*dt*f3/m")
        self.addConstrainVelocities()
        # Compute final total energy
        self.addComputeSum("ke", "0.5*m*v*v")
        self.addComputeGlobal("potential_new", "energy")
        self.addComputeGlobal("Enew", "ke + potential_new")
        # Accept/reject, ensuring rejection if energy is NaN
        self.addComputeGlobal("accept", "step(exp(-(Enew-Eold)/kT) - uniform)")
        self.beginIfBlock("accept != 1")
        self.addComputePerDof("x", "xold")
        self.addComputePerDof("v", "-vold")
        self.addComputeGlobal("potential_new", "potential_old")
        self.endBlock()

        #
        # Velocity randomization
        #
        self.addComputePerDof("v", "sqrt(b)*v + sqrt(1-b)*sigma*gaussian")
        self.addConstrainVelocities()

        #
        # Accumulate statistics.
        #
        self.addComputeGlobal("naccept", "naccept + accept")
        self.addComputeGlobal("ntrials", "ntrials + 1")
    
    def resetStatistics(self):
        """
        Reset the step counter and statistics

        """
        self.setGlobalVariableByName('ntrials', 0)
        self.setGlobalVariableByName('naccept', 0)

    def setTemperature(self, temperature):
        """
        Set the temperature.

        This also resets the trial statistics.
        
        Parameters
        ----------
        temperature : unit.Quantity
            The new temperature
        """
        kT = kB * temperature
        self.setGlobalVariableByName('kT', kT)
        # Reset statistics to ensure 'sigma' is updated on step 0
        self.resetStatistics()


temperature = 300*unit.kelvin

platform = openmm.Platform.getPlatformByName('CUDA')
try:
    platform.setPropertyDefaultValue('Precision', 'mixed')
    platform.setPropertyDefaultValue('DeterministicForces', 'true')
except:
    pass
#platform = openmm.Platform.getPlatformByName('CPU')

switch_width = 1.0 * unit.angstroms
use_dispersion_correction = True
testsystem_options = ['SrcExplicit', 'DHFRExplicit']
use_switching_function_options = [True]
constraint_tol_options = np.array([1.0e-8])
ewald_tol_options = [1.0e-4, 1.0e-5]
cutoff_options = np.array([10.0]) * unit.angstroms
hmr_options = [True]
timestep_options = np.array([2.0, 1.5, 1.0]) * unit.femtoseconds
nvalence_options = [1, 2, 4]
nshort_options = [1, 2, 4]
nmedium_options = [1, 2, 4]
solute_solvent_splitting_options = [False, True]

options_list = [solute_solvent_splitting_options, hmr_options, nvalence_options, nshort_options, nmedium_options, timestep_options, constraint_tol_options, ewald_tol_options, cutoff_options, use_switching_function_options, testsystem_options]

try:
    print('Importing MPI...')
    from mpi4py import MPI
    rank = MPI.COMM_WORLD.rank
    size = MPI.COMM_WORLD.size
except Exception as e:
    print(e)
    print('Could not start MPI.')
    rank = 0
    size = 1

#print('Configuring logger...')
#from yank.utils import config_root_logger
#verbose = True
#config_root_logger(verbose, mpicomm=mpicomm)

def select_options(options_list, index):
    selected_options = list()
    for option in options_list:
        noptions = len(option)
        selected_options.append(option[index % noptions])
        index = int(index/noptions)
    return selected_options

noptionsets = 1
for optionset in options_list:
    noptionsets *= len(optionset)
if rank == 0: print('There are %d optionsets to try' % noptionsets)

def remove_CMMotionRemover(system):
    # Remove CMMotionRemover
    forces_to_remove = list()
    for (index, force) in enumerate(system.getForces()):
        if force.__class__.__name__ in ['CMMotionRemover']:
            forces_to_remove.append(index)
    for index in forces_to_remove[::-1]:
        system.removeForce(index)

# Code for creating direct space mimic
def setup_xu_mts(system, topology=None, solute_solvent_splitting=False):
    """
    Set up force groups for an MTS scheme following:

    Zhou, Harder, Xu, Berne. 
    Efficient multiple time step method for use with Ewald and particle mesh Ewald for large biomolecular systems.
    JCP 115:2348, 2001.
    http://dx.doi.org/10.1063/1.1385159

    force groups:
    0: valence terms
    1: NonbondedForce 0 <= r <= 7  with soft switches
    2: NonbondedForce 6 <= r <= 10 with soft switches
    3: long-range: NonbondedForce, subtracting force groups 1 and 2

    Currently, this only works for one NonbondedForce in the original system

    Parameters
    ----------
    system : simtk.openmm.System
       The original system, which will be modified in place.
    topology : simtk.openmm.app.Topology, optional, default=None
       Topology used for determining solute-solvent splitting
    solute_solvent_splitting : bool
       If True, will put solute-solute and solute-solvent terms in force groups 0,1,2
       while remaining forces go into force group 3.

    """
    # Determine solvent atoms
    solute_atoms = set()
    solvent_atoms = set()
    for atom in topology.atoms():
        if atom.residue.name in ['HOH', 'WAT']:
            solvent_atoms.add(atom.index)
        else:
            solute_atoms.add(atom.index)
    all_atoms = solute_atoms.union(solvent_atoms)

    # Reference NonbondedForce info
    forces = { force.__class__.__name__ : force for force in system.getForces() }
    reference_force = forces['NonbondedForce']
    ONE_4PI_EPS0 = 138.935456
    [alpha_ewald, nx, ny, nz] = reference_force.getPMEParameters()
    if (alpha_ewald/alpha_ewald.unit) == 0.0:
        # If alpha is 0.0, alpha_ewald is computed by OpenMM from from the error tolerance.
        tol = reference_force.getEwaldErrorTolerance()
        alpha_ewald = (1.0/reference_force.getCutoffDistance()) * np.sqrt(-np.log(2.0*tol))

    # Force group 0: Valence terms and exceptions
    for force in system.getForces():
        if force.__class__.__name__ != 'NonbondedForce':
            force.setForceGroup(0)

    expression = '4*epsilon*((sigma/r)^12 - (sigma/r)^6) + ONE_4PI_EPS0*chargeprod/r;'
    expression += 'ONE_4PI_EPS0 = %.16e;' % (ONE_4PI_EPS0)
    force = openmm.CustomBondForce(expression)
    force.addPerBondParameter("chargeprod")
    force.addPerBondParameter("sigma")
    force.addPerBondParameter("epsilon")
    for index in range(reference_force.getNumExceptions()):
        [iatom, jatom, chargeprod, sigma, epsilon] = reference_force.getExceptionParameters(index)
        if (chargeprod/chargeprod.unit != 0.0) or (epsilon/epsilon.unit != 0.0):
            force.addBond(iatom, jatom, [chargeprod, sigma, epsilon])
        # Disable exceptions in NonbondedForce since we've moved them to inner timestep
        reference_force.setExceptionParameters(index, iatom, jatom, chargeprod*0, sigma, epsilon*0)
    force.setForceGroup(0)
    system.addForce(force)

    # Force group 1: 0 <= r <= 7
    expression = 'S0*(4*epsilon*((sigma/r)^12 - (sigma/r)^6) + ONE_4PI_EPS0*chargeprod/r);'
    expression += 'chargeprod = charge1*charge2;'
    expression += 'epsilon = sqrt(epsilon1*epsilon2);'
    expression += 'sigma = 0.5*(sigma1+sigma2);'
    expression += 'ONE_4PI_EPS0 = %.16e;' % (ONE_4PI_EPS0)
    expression += 'S0 = step(r0switch-r) + step(r-r0switch)*step(r0cutoff-r)*(1 - 6*x0^5 + 15*x0^4 - 10*x0^3); x0 = (r-r0switch)/(r0cutoff-r0switch); r0switch=%f; r0cutoff=%f;' % (0.6, 0.7)
    force = openmm.CustomNonbondedForce(expression)
    cutoff_distance = 7.0 * unit.angstroms
    switch_distance = 1.0 * unit.angstroms
    force.addPerParticleParameter("charge")
    force.addPerParticleParameter("sigma")
    force.addPerParticleParameter("epsilon")
    force.setUseSwitchingFunction(False)
    force.setNonbondedMethod(openmm.CustomNonbondedForce.CutoffPeriodic)
    force.setCutoffDistance(cutoff_distance)
    force.setSwitchingDistance(switch_distance)
    force.setUseLongRangeCorrection(False)
    for index in range(reference_force.getNumParticles()):
        [charge, sigma, epsilon] = reference_force.getParticleParameters(index)
        force.addParticle([charge, sigma, epsilon])
    for index in range(reference_force.getNumExceptions()):
        [iatom, jatom, chargeprod, sigma, epsilon] = reference_force.getExceptionParameters(index)
        force.addExclusion(iatom, jatom)
    force.setForceGroup(1)
    if solute_solvent_splitting:
        force.addInteractionGroup(solute_atoms, all_atoms)
    system.addForce(force)
    
    # Force group 2: 6 <= r <= 10
    expression = '(1-S0)*S1*(4*epsilon*((sigma/r)^12 - (sigma/r)^6) + ONE_4PI_EPS0*chargeprod/r);'
    expression += 'chargeprod = charge1*charge2;'
    expression += 'epsilon = sqrt(epsilon1*epsilon2);'
    expression += 'sigma = 0.5*(sigma1+sigma2);'
    expression += 'ONE_4PI_EPS0 = %.16e;' % (ONE_4PI_EPS0)
    expression += 'S0 = step(r0switch-r) + step(r-r0switch)*step(r0cutoff-r)*(1 - 6*x0^5 + 15*x0^4 - 10*x0^3); x0 = (r-r0switch)/(r0cutoff-r0switch); r0switch=%f; r0cutoff=%f;' % (0.6, 0.7)
    expression += 'S1 = step(r1switch-r) + step(r-r1switch)*step(r1cutoff-r)*(1 - 6*x1^5 + 15*x1^4 - 10*x1^3); x1 = (r-r1switch)/(r1cutoff-r1switch); r1switch=%f; r1cutoff=%f;' % (0.9, 1.0)
    force = openmm.CustomNonbondedForce(expression)
    cutoff_distance = 10.0 * unit.angstroms
    switch_distance = 1.0 * unit.angstroms
    force.addPerParticleParameter("charge")
    force.addPerParticleParameter("sigma")
    force.addPerParticleParameter("epsilon")
    force.setUseSwitchingFunction(False)
    force.setNonbondedMethod(openmm.CustomNonbondedForce.CutoffPeriodic)
    force.setCutoffDistance(cutoff_distance)
    force.setSwitchingDistance(switch_distance)
    force.setUseLongRangeCorrection(False)
    for index in range(reference_force.getNumParticles()):
        [charge, sigma, epsilon] = reference_force.getParticleParameters(index)
        force.addParticle([charge, sigma, epsilon])
    for index in range(reference_force.getNumExceptions()):
        [iatom, jatom, chargeprod, sigma, epsilon] = reference_force.getExceptionParameters(index)
        force.addExclusion(iatom, jatom)
    force.setForceGroup(2)
    if solute_solvent_splitting:
        force.addInteractionGroup(solute_atoms, all_atoms)
    system.addForce(force)

    # Force group 3: Long-range
    cutoff_distance = 10.0 * unit.angstroms
    switch_distance = 1.0 * unit.angstroms
    reference_force.setUseSwitchingFunction(True)
    reference_force.setNonbondedMethod(openmm.NonbondedForce.PME)
    reference_force.setCutoffDistance(cutoff_distance)
    force.setSwitchingDistance(switch_distance)
    reference_force.setReciprocalSpaceForceGroup(3)
    reference_force.setForceGroup(3)
    # Subtract off short-range terms (force groups 1+2)
    expression = '-1*S1*(4*epsilon*((sigma/r)^12 - (sigma/r)^6) + ONE_4PI_EPS0*chargeprod/r);'
    expression += 'chargeprod = charge1*charge2;'
    expression += 'epsilon = sqrt(epsilon1*epsilon2);'
    expression += 'sigma = 0.5*(sigma1+sigma2);'
    expression += 'ONE_4PI_EPS0 = %.16e;' % (ONE_4PI_EPS0)
    expression += 'S1 = step(r1switch-r) + step(r-r1switch)*step(r1cutoff-r)*(1 - 6*x1^5 + 15*x1^4 - 10*x1^3); x1 = (r-r1switch)/(r1cutoff-r1switch); r1switch=%f; r1cutoff=%f;' % (0.9, 1.0)
    force = openmm.CustomNonbondedForce(expression)
    force.addPerParticleParameter("charge")
    force.addPerParticleParameter("sigma")
    force.addPerParticleParameter("epsilon")
    force.setUseSwitchingFunction(False)
    force.setNonbondedMethod(openmm.CustomNonbondedForce.CutoffPeriodic)
    force.setCutoffDistance(cutoff_distance)
    force.setSwitchingDistance(switch_distance)
    force.setUseLongRangeCorrection(False)
    for index in range(reference_force.getNumParticles()):
        [charge, sigma, epsilon] = reference_force.getParticleParameters(index)
        force.addParticle([charge, sigma, epsilon])
    for index in range(reference_force.getNumExceptions()):
        [iatom, jatom, chargeprod, sigma, epsilon] = reference_force.getExceptionParameters(index)
        force.addExclusion(iatom, jatom)
    force.setForceGroup(3)
    if solute_solvent_splitting:
        force.addInteractionGroup(solute_atoms, all_atoms)
    system.addForce(force)

# Set up test systems
testsystem_library = dict()
for testsystem_name in testsystem_options:
    from openmmtools import testsystems
    testsystem = getattr(testsystems, testsystem_name)()
    remove_CMMotionRemover(testsystem.system)
    testsystem_library[testsystem_name] = testsystem

for index in range(rank, noptionsets, size):
    [solute_solvent_splitting, hmr, nvalence, nshort, nmedium, timestep, constraint_tol, ewald_tol, cutoff, use_switching_function, testsystem_name] = select_options(options_list, index)

    testsystem = testsystem_library[testsystem_name]

    # Deep copy system and set periodic cutoff
    system = copy.deepcopy(testsystem.system)
    for force in system.getForces():
        if hasattr(force, 'setCutoffDistance'):
            force.setCutoffDistance(cutoff)
        if hasattr(force, 'setUseSwitchingFunction'):
            force.setUseSwitchingFunction(use_switching_function)
        if hasattr(force, 'setSwitchingDistance'):
            switch_distance = cutoff-switch_width
            force.setSwitchingDistance(switch_distance)
        if hasattr(force, 'setEwaldErrorTolerance'):
            force.setEwaldErrorTolerance(ewald_tol)
        if hasattr(force, 'setUseDispersionCorrection'):
            force.setUseDispersionCorrection(use_dispersion_correction)

    if hmr:
        # Set heavy hydrogen masses if hmr option is set.
        # Note: This does not currently subtract mass from their bonded partners.
        for (atom_index, atom) in enumerate(testsystem.topology.atoms()):
            if atom.element.symbol == "H":
                system.setParticleMass(atom_index, 4.0)

    # Set up Xu MTS splitting
    setup_xu_mts(system, testsystem.topology, solute_solvent_splitting=solute_solvent_splitting)

    # Equilibrate a bit
    integrator = MTSGHMCIntegrator(temperature, 5.0/unit.picoseconds, 0.2 * unit.femtoseconds)
    context = openmm.Context(system, integrator, platform)
    context.setPositions(testsystem.positions)
    #context.setVelocitiesToTemperature(temperature)
    integrator.step(nsteps)
    positions = context.getState(getPositions=True).getPositions(asNumpy=True)
    naccepted = integrator.getGlobalVariableByName('naccept')
    nattempted = integrator.getGlobalVariableByName('ntrials')
    del context, integrator

    # Create integrator
    integrator = MTSGHMCIntegrator(temperature, 5.0/unit.picoseconds, timestep, nvalence=nvalence, nshort=nshort, nmedium=nmedium)
    integrator.setConstraintTolerance(constraint_tol)
        
    # Create Context
    context = openmm.Context(system, integrator, platform)
    context.setPositions(positions)
    context.setVelocitiesToTemperature(temperature)

    # Warm up integrator
    integrator.step(nsteps)
    integrator.resetStatistics()

    # Run dynamics.
    initial_time = time.time()
    integrator.step(nsteps)
    final_time = time.time()

    naccepted = integrator.getGlobalVariableByName('naccept')
    nattempted = integrator.getGlobalVariableByName('ntrials')

    # Clean up.
    del context, integrator

    elapsed_time = final_time - initial_time
    ns_per_day = 24. * 60. * 60. / elapsed_time * nsteps * timestep / unit.nanoseconds
    print('%-24s | solute_solvent_splitting %5s | constraint_tol %8.3e | ewald_tol %8.3e | cutoff %8.3f A | HMR %8s | timestep %8.3f fs | nvalence %2d | nshort %2d | nmedium %2d | %8.1f %% | %8.1f ns/day' % (testsystem_name, str(solute_solvent_splitting), constraint_tol, ewald_tol, cutoff/unit.angstroms, hmr, timestep/unit.femtoseconds, nvalence, nshort, nmedium, float(naccepted)/float(nattempted)*100, ns_per_day))
    sys.stdout.flush()
```

**Comment by peastman** [comment]
The really big difference is that interaction groups don't use the neighbor list.  It just computes every interaction between the two sets.  That's fine for small groups, but when you're looking at 5000 atoms interacting with 45,000 atoms, it's more expensive than computing everything.

In theory we could create a new neighbor list structure specifically for the interaction groups.  I'm not sure how that would work, and it would probably be complicated to implement, but it might be possible.

**Comment by jchodera** [comment]
> The really big difference is that interaction groups don't use the neighbor list. It just computes every interaction between the two sets. That's fine for small groups, but when you're looking at 5000 atoms interacting with 45,000 atoms, it's more expensive than computing everything.

Thanks! That definitely explains what we're seeing.

> In theory we could create a new neighbor list structure specifically for the interaction groups. I'm not sure how that would work, and it would probably be complicated to implement, but it might be possible.

If this was possible, it would mean that these inner timestep force evaluations involving only solute-solute and solute-solvent interactions could be evaluated at greatly reduced cost (potentially 5x-10x less?).

Could a modified form of [`findBlocksWithInteractions()`](https://github.com/pandegroup/openmm/blob/master/platforms/cuda/src/kernels/findInteractingBlocks.cu#L129-L178) be used here, where interaction groups are used to identify excluded sets of atoms instead of the exclusion lists? 

Alternatively, @pgrinaway suggests that the standard pairlist scheme could be used and a check of whether any interactions in the tiles are permitted by interaction groups added to see if the interactions should be computed.

I imagine that the pairlist-based scheme might be *less* efficient than the current scheme where all interactions are computed for small sets of atoms, so there might be some logic that automatically chooses which scheme should be used (or lets the user select).

**Comment by jchodera** [comment]
Closing this since neighbor lists were implemented in #2014!
